File: GroceryListHelper.Client/Services/StoreProductsAPIService.cs
namespace GroceryListHelper.Client.Services;

public sealed class StoreProductsAPIService(IHttpClientFactory clientFactory) : IStoreProductsService
{
    private readonly HttpClient client = clientFactory.CreateClient("Client");
    private const string uri = "api/storeproducts";

    public async Task<List<StoreProduct>> GetStoreProducts()
    {
        return await client.GetFromJsonAsync<List<StoreProduct>>(uri) ?? [];
    }

    public async Task SaveStoreProduct(StoreProduct product)
    {
        HttpResponseMessage response = await client.PostAsJsonAsync(uri, product);
        response.EnsureSuccessStatusCode();
    }

    public async Task ClearStoreProducts()
    {
        HttpResponseMessage response = await client.DeleteAsync(uri);
        response.EnsureSuccessStatusCode();
    }

    public async Task UpdateStoreProduct(StoreProduct storeProduct)
    {
        HttpResponseMessage response = await client.PutAsJsonAsync(uri, storeProduct);
        response.EnsureSuccessStatusCode();
    }


}

File: GroceryListHelper.Client/src/services/StoreProductsApiService.ts
import type { StoreProduct } from "../types/StoreProduct";
import type { IStoreProductsService } from "../types/IStoreProductsService";
import { get, send } from "../helpers/httpHelpers";

export class StoreProductsApiService implements IStoreProductsService {
    readonly url = "api/storeproducts/";

    getStoreProducts = async () => {
        const productsResponse = await get(this.url, "GET", async (r) => await r.json() as StoreProduct[]);
        return productsResponse;
    }

    createStoreProduct = async (product: StoreProduct) => {
        const response = await send(this.url, "POST", product);
        return response;
    }

    updateStoreProduct = async (product: StoreProduct) => {
        const response = await send(this.url, "POST", product);
        return response;
    }

    deleteStoreProducts = async () => {
        const response = await send(this.url, "DELETE");
        return response;
    }
}